File: a4/main.py
with open("input.txt", "r") as f:
	assignments = [line.strip().split(",") for line in f.readlines()]
	assignments = [(pair[0].split("-"), pair[1].split("-")) for pair in assignments]
	assignments = [(tuple(map(int, pair[0])), tuple(map(int, pair[1]))) for pair in assignments]

def contains(a, b):
	return a[0] <= b[0] and a[1] >= b[1]

def overlap(a, b):
	return contains(a, b) or (a[0] <= b[1] and a[1] >= b[1])

def solve1(assignments):
	return len(list(filter(lambda x: contains(x[0], x[1]) or contains(x[1], x[0]), assignments)))

def solve2(assignments):
	return len(list(filter(lambda x: overlap(x[0], x[1]) or overlap(x[1], x[0]), assignments)))

print(solve1(assignments))
print(solve2(assignments))

File: a4/main.rs
use std::fs;

fn contains((a, b): &(i32, i32), (c, d): &(i32, i32)) -> bool {
	a <= c && b >= d
}
fn overlap((a, b): &(i32, i32), (c, d): &(i32, i32)) -> bool {
	contains(&(*a, *b), &(*c, *d)) || (a <= d && b >= d)
}

fn solve1(assignments: &[((i32, i32), (i32, i32))]) -> i32 {
	assignments.iter().filter(|(a, b)| contains(a, b) || contains(b, a)).count() as i32
}

fn solve2(assignments: &[((i32, i32), (i32, i32))]) -> i32 {
	assignments.iter().filter(|(a, b)| overlap(a, b) || overlap(b, a)).count() as i32
}

fn main() {
	let assignments: Vec<((i32, i32), (i32, i32))> = fs::read_to_string("input.txt")
		.expect("File does not exist")
		.split("\n")
		.map(|a| {
			let mut q = a.splitn(2, ",")
				.map(|b| {
					let mut q = b.splitn(2, "-")
						.map(|s| s.parse::<i32>().unwrap());
					(q.next().unwrap(), q.next().unwrap())
				});
			(q.next().unwrap(), q.next().unwrap())
		}).collect();
	println!("{:?}", solve1(&assignments));
	println!("{:?}", solve2(&assignments));
}
